python/reactivedataflow/reactivedataflow/execution_graph.py

```diff
@@ -16,20 +16,35 @@ class ExecutionGraph:
     _nodes: dict[str, Node]
     _outputs: dict[str, Output]
     _order: list[str]
+    _on_node_start: OnNodeStartCallback
+    _on_node_finish: OnNodeFinishCallback
 
     def __init__(
-        self, nodes: dict[str, Node], outputs: dict[str, Output], order: list[str]
+        self,
+        nodes: dict[str, Node],
+        outputs: dict[str, Output],
+        order: list[str],
+        on_node_start: OnNodeStartCallback | None = None,
+        on_node_finish: OnNodeFinishCallback | None = None,
     ):
         """Initialize the execution graph.
 
         Args:
             nodes: The nodes in the graph.
             outputs: The outputs of the graph.
             order: The topological order of the nodes, starting with input nodes.
+            on_node_start: The callback for when a node starts.
```

Comment:
One thing that doesn't clicks much with me is how much code is required to modify callbacks, or to add a new one if needed. Will sound cliche, but, have you tried with Protocols?

Reply:
Yeah I think you have a point here. There are a few layers that these have to be wired through. I replaced individual callbacks with a top-level Callbacks protocol.